### DISCUSSION TITLE
Generic Mapper for ValueTypes

### DISCUSSION THREAD

**Discussion by nbeDPDHL** [mapstruct/mapstruct#3108] [Q&A]
We have a lot of ValueTypes (i.e. Simple Wrapper Classes around one property of a very simple type (basically int, String, UUID, etc. for which mappers exist)).

They basically always look like this

```
public record AnyName(String value) {
    public AnyName(String value) {
        Objects.requireNonNull(value);
        this.value = valuet;
    }

    public String value() {
        return this.value;
    }
}
```

(this is, of course, the example for a String ValueType only)

Now i don't want to write a Mapper (or declare a Mapper interface) like this

```
public interface AnyNameMapper {

    default String map(AnyName x) {
        return x == null ? null : x.value();
    }

    @Mapping(target = "value", source = ".")
    AnyName map(String value);
}
```

for every single ValueType class, but rather want to declare this in a very generic way, such that it will be applied to any ValueType, that just wraps a single field (of name "value" (and of type String, in this example)).

Is that possible?
I would be content if I had to write such a generic mapper for every single basic type that can occur (instead of using generics), i.e. one for ValueType\<String\>, one for ValueType\<UUID\>, etc.. But right now I don't see how I can achieve this.

**Comment by dkellenb** [comment]
We have the same thing in our different projects, but IMO you can just define them in a `CoreMapper` and inject this `CoreMapper` in all other mappers. Indeed, it's repeating work, but you do it only once.

At least for all common toString, toUuid, ... mappings, you only need to define it once: Define an interface `StringWrapper` with a method `toString()`, all string wrappers can implement this interface and define one mapper like this: `default String map(StringWrapper sw) { return sw == null ? null : sw.getString(); } `. But for constructing such wrapper objects, there is AFAIK no shortcut.

But indeed, if _MapStruct_ would check if there is an appropriate constructor available, it could just take the constructor to create it. Example:

```java
@Data
class CustomerId {
  private final UUID uuid;
}
```

If the source is a `UUID` and the destination a `CustomerId`, _MapStruct_ may check if there is also a Constructor on `CustomerId` accepting a `UUID`.
